registrar/flow.py

```diff
@@ -0,0 +1,104 @@
+import json
+import logging
+import os
+from functools import wraps
+
+import yaml
+from dask_kubernetes.objects import make_pod_spec
+from gcsfs import GCSFileSystem
+from pangeo_forge_recipes.recipes.base import BaseRecipe
+from pangeo_forge_recipes.storage import CacheFSSpecTarget, MetadataTarget
+from prefect.executors.dask import DaskExecutor
+from prefect.run_configs.kubernetes import KubernetesRun
+from prefect.storage import GCS
+from s3fs import S3FileSystem
+
+REQUIRED_ENV_VARS = [
+    "PROJECT_NAME",  # bakery project
+    "GOOGLE_APPLICATION_CREDENTIALS",  # path to json creds for bakery project
+    "STORAGE_NAME",  # bakery project bucket name
+    "PANGEO_FORGE_OSN_KEY",
+    "PANGEO_FORGE_OSN_SECRET",
+    "BAKERY_IMAGE",  # Seems like this is how we'd set worker images dynamically?
+    "PREFECT_PROJECT",
+    "PREFECT__CLOUD__AUTH_TOKEN",
+    "PREFECT__CLOUD__AGENT__LABELS",
+]
+for v in REQUIRED_ENV_VARS:
+    if v not in os.environ:
+        raise ValueError(f"Environment variable {v} not set. Required for flow registration.")
+
+
+def set_log_level(func):
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        logging.basicConfig()
+        logging.getLogger("pangeo_forge_recipes").setLevel(level=logging.DEBUG)
+        result = func(*args, **kwargs)
+        return result
+
+    return wrapper
+
+
+def register_flow(recipe: BaseRecipe, flow_name: str):
+    storage_name = os.environ["STORAGE_NAME"]
+    fs_gcs = GCSFileSystem(project=os.environ["PROJECT_NAME"], bucket=storage_name)
+    fs_osn = S3FileSystem(
+        key=os.environ["PANGEO_FORGE_OSN_KEY"],
+        secret=os.environ["PANGEO_FORGE_OSN_SECRET"],
+        client_kwargs=dict(endpoint_url="https://ncsa.osn.xsede.org"),
+        default_cache_type="none",
+        default_fill_cache=False,
+        use_listings_cache=False,
+    )
+    target_base = "s3://Pangeo/pangeo-forge"
+    recipe.target = MetadataTarget(fs_osn, root_path=f"{target_base}/pfcsb-test/{flow_name}")
```

Comment:
```suggestion
    recipe.target = FSSpecTarget(fs_osn, root_path=f"{target_base}/pfcsb-test/{flow_name}.zarr")
```

Also, we should be using the recipe name / id, as [specified in ARD 2](https://github.com/pangeo-forge/roadmap/blob/master/doc/adr/0002-use-meta-yaml-to-track-feedstock-metadata.md#top-level-data) as the name. See also https://github.com/pangeo-forge/roadmap/blob/master/doc/adr/0003-standardize-storage-target-layout.md

Reply:
> Also, we should be using the recipe name / id, as [specified in ARD 2](https://github.com/pangeo-forge/roadmap/blob/master/doc/adr/0002-use-meta-yaml-to-track-feedstock-metadata.md#top-level-data) as the name.

✔️ https://github.com/pangeo-forge/registrar/pull/1/commits/63a295487698cfb7b22e6c5174ce83748e5ed1b1

> See also https://github.com/pangeo-forge/roadmap/blob/master/doc/adr/0003-standardize-storage-target-layout.md

This ADR says we should include `feedstock_name` in the path. Currently, our database doesn't store this information exactly; instead, we have a `Feedstock.spec`, which might be, e.g., [`"pangeo-forge/noaa-oisst-avhrr-only-feedstock"`](https://api-staging.pangeo-forge.org/feedstocks/1). Including this whole string seems redundant (because "pangeo-forge" is captured elsewhere in the path)... I wonder if we maybe don't need to encode this information in the paths anymore, now that we can link datasets back to feedstocks via the database?